File: sdk/go/transport.go
package velocity_mcp

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"sync"
	"time"

	"github.com/gorilla/websocket"
)

// Transport is the interface for MCP transports
type Transport interface {
	Send(ctx context.Context, request JsonRpcRequest) (*JsonRpcResponse, error)
	Close() error
}

// HttpTransport implements HTTP transport
type HttpTransport struct {
	url     string
	apiKey  string
	client  *http.Client
	timeout time.Duration
}

// NewHttpTransport creates a new HTTP transport
func NewHttpTransport(url string, apiKey string, timeout time.Duration) *HttpTransport {
	if timeout == 0 {
		timeout = 30 * time.Second
	}
	return &HttpTransport{
		url:     url,
		apiKey:  apiKey,
		client:  &http.Client{Timeout: timeout},
		timeout: timeout,
	}
}

// Send sends a JSON-RPC request over HTTP
func (t *HttpTransport) Send(ctx context.Context, request JsonRpcRequest) (*JsonRpcResponse, error) {
	body, err := json.Marshal(request)
	if err != nil {
		return nil, &TransportError{McpError{Message: fmt.Sprintf("failed to marshal request: %v", err)}}
	}

	req, err := http.NewRequestWithContext(ctx, "POST", t.url, bytes.NewReader(body))
	if err != nil {
		return nil, &TransportError{McpError{Message: fmt.Sprintf("failed to create request: %v", err)}}
	}

	req.Header.Set("Content-Type", "application/json")
	if t.apiKey != "" {
		req.Header.Set("Authorization", "Bearer "+t.apiKey)
	}

	resp, err := t.client.Do(req)
	if err != nil {
		return nil, &TransportError{McpError{Message: fmt.Sprintf("failed to send request: %v", err)}}
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, &TransportError{McpError{Message: fmt.Sprintf("HTTP error %d: %s", resp.StatusCode, string(body))}}
	}

	var rpcResp JsonRpcResponse
	if err := json.NewDecoder(resp.Body).Decode(&rpcResp); err != nil {
		return nil, &TransportError{McpError{Message: fmt.Sprintf("failed to decode response: %v", err)}}
	}

	return &rpcResp, nil
}

// Close closes the HTTP transport
func (t *HttpTransport) Close() error {
	return nil
}

// WebSocketTransport implements WebSocket transport
type WebSocketTransport struct {
	url    string
	apiKey string
	conn   *websocket.Conn
	mu     sync.Mutex
}

// NewWebSocketTransport creates a new WebSocket transport
func NewWebSocketTransport(url string, apiKey string) *WebSocketTransport {
	return &WebSocketTransport{
		url:    url,
		apiKey: apiKey,
	}
}

// Send sends a JSON-RPC request over WebSocket
func (t *WebSocketTransport) Send(ctx context.Context, request JsonRpcRequest) (*JsonRpcResponse, error) {
	t.mu.Lock()
	defer t.mu.Unlock()

	// Connect if not connected
	if t.conn == nil {
		header := http.Header{}
		if t.apiKey != "" {
			header.Set("Authorization", "Bearer "+t.apiKey)
		}
		
		conn, _, err := websocket.DefaultDialer.DialContext(ctx, t.url, header)
		if err != nil {
			return nil, &TransportError{McpError{Message: fmt.Sprintf("failed to connect: %v", err)}}
		}
		t.conn = conn
	}

	// Send request
	body, err := json.Marshal(request)
	if err != nil {
		return nil, &TransportError{McpError{Message: fmt.Sprintf("failed to marshal request: %v", err)}}
	}

	if err := t.conn.WriteMessage(websocket.TextMessage, body); err != nil {
		return nil, &TransportError{McpError{Message: fmt.Sprintf("failed to send request: %v", err)}}
	}

	// Receive response
	_, message, err := t.conn.ReadMessage()
	if err != nil {
		return nil, &TransportError{McpError{Message: fmt.Sprintf("failed to receive response: %v", err)}}
	}

	var rpcResp JsonRpcResponse
	if err := json.Unmarshal(message, &rpcResp); err != nil {
		return nil, &TransportError{McpError{Message: fmt.Sprintf("failed to decode response: %v", err)}}
	}

	return &rpcResp, nil
}

// Close closes the WebSocket transport
func (t *WebSocketTransport) Close() error {
	t.mu.Lock()
	defer t.mu.Unlock()
	
	if t.conn != nil {
		return t.conn.Close()
	}
	return nil
}

File: sdk/typescript/src/transport.ts
/**
 * Transport implementations for MCP client
 */

import axios, { AxiosInstance } from "axios";
import { spawn, ChildProcess } from "child_process";
import { JsonRpcRequest, JsonRpcResponse } from "./types";
import { TransportError, ConnectionClosedError } from "./errors";

export interface Transport {
  send(request: JsonRpcRequest): Promise<JsonRpcResponse>;
  close(): Promise<void>;
}

export interface HttpTransportOptions {
  apiKey?: string;
  timeout?: number;
}

export class HttpTransport implements Transport {
  private client: AxiosInstance;
  private url: string;

  constructor(url: string, options: HttpTransportOptions = {}) {
    this.url = url;
    const { apiKey, timeout = 30000 } = options;

    this.client = axios.create({
      timeout,
      headers: {
        "Content-Type": "application/json",
        ...(apiKey && { Authorization: `Bearer ${apiKey}` }),
      },
    });
  }

  async send(request: JsonRpcRequest): Promise<JsonRpcResponse> {
    try {
      const response = await this.client.post(this.url, request);
      return response.data;
    } catch (error: any) {
      if (error.code === "ECONNABORTED") {
        throw new TransportError("Request timed out");
      }
      if (error.response) {
        throw new TransportError(
          `HTTP error ${error.response.status}: ${error.response.data}`
        );
      }
      throw new TransportError(`Transport error: ${error.message}`);
    }
  }

  async close(): Promise<void> {
    // HTTP transport doesn't need explicit close
  }
}

export interface StdioTransportOptions {
  command: string;
  args?: string[];
}

export class StdioTransport implements Transport {
  private process: ChildProcess | null = null;
  private command: string;
  private args: string[];
  private requestId = 0;

  constructor(options: StdioTransportOptions) {
    this.command = options.command;
    this.args = options.args || [];
  }

  private ensureProcess(): ChildProcess {
    if (!this.process || this.process.exitCode !== null) {
      this.process = spawn(this.command, this.args, {
        stdio: ["pipe", "pipe", "pipe"],
      });

      this.process.on("error", (error) => {
        throw new TransportError(`Process error: ${error.message}`);
      });
    }
    return this.process;
  }

  async send(request: JsonRpcRequest): Promise<JsonRpcResponse> {
    const process = this.ensureProcess();

    if (!process.stdin || !process.stdout) {
      throw new ConnectionClosedError("Process stdin/stdout not available");
    }

    return new Promise((resolve, reject) => {
      const requestJson = JSON.stringify(request);
      
      process.stdin!.write(requestJson + "\n");

      let responseData = "";
      const onData = (data: Buffer) => {
        responseData += data.toString();
        if (responseData.includes("\n")) {
          process.stdout!.removeListener("data", onData);
          try {
            const response = JSON.parse(responseData.trim());
            resolve(response);
          } catch (error: any) {
            reject(new TransportError(`Failed to parse response: ${error.message}`));
          }
        }
      };

      process.stdout.on("data", onData);

      process.once("exit", (code) => {
        if (code !== null && code !== 0) {
          reject(new ConnectionClosedError(`Process exited with code ${code}`));
        }
      });
    });
  }

  async close(): Promise<void> {
    if (this.process && this.process.exitCode === null) {
      this.process.kill();
      await new Promise<void>((resolve) => {
        this.process!.once("exit", () => resolve());
      });
    }
  }
}
